torch/csrc/distributed/rpc/python_rpc_handler.cpp

```diff
@@ -68,27 +68,33 @@ void cleanupPyObj(py::object& obj) {
 
 } // namespace
 
-PythonRpcHandler::PythonRpcHandler() {
-  PROFILE_GIL_SCOPED_ACQUIRE;
-  py::object rpcInternal = py::module::import(kInternalModule);
-  py::object rpcApi = py::module::import("torch.distributed.rpc.api");
-  py::object rrefProxy = py::module::import("torch.distributed.rpc.rref_proxy");
-
-  pyRunFunction_ = getFunction(rpcInternal, "_run_function");
-  pySerialize_ = getFunction(rpcInternal, "serialize");
-  pyDeserialize_ = getFunction(rpcInternal, "deserialize");
-  pyHandleException_ = getFunction(rpcInternal, "_handle_exception");
-
-  rrefProxyFunctions_.rpcSync_ = getFunction(rpcApi, "rpc_sync");
-  rrefProxyFunctions_.rpcAsync_ = getFunction(rpcApi, "rpc_async");
-  rrefProxyFunctions_.remote_ = getFunction(rpcApi, "remote");
-  rrefProxyFunctions_.rrefProxyCtor_ = getFunction(rrefProxy, "RRefProxy");
-
-  jitCompilationUnit_ = torch::jit::get_python_cu();
-  typeParser_ = std::make_shared<jit::ScriptTypeParser>(
-      std::make_shared<PythonTypeResolver>());
+void PythonRpcHandler::init() {
+  if (!initialized_) {
```

Comment:
I'm wondering what's the motivation for this atomic and if it's to prevent races? If there could be thread races, this atomic won't be enough to ensure initialization only once right? If there's no races, why do we need an atomic?

Reply:
This function can probably be executed twice and that is safe and would happen rarely. Apart from that, the atomic is mostly to act as a memory barrier to ensure other threads are informed about this change. For example, if one thread calls `cleanup` and another thread calls `getInstance` after it, if we don't use an atomic the thread calling `getInstance` might still see the old value of `initialized`.